File: sdk/elixir/lib/bsdkrun.ex
defmodule Bsdkrun do
  @moduledoc """
  Elixir SDK for [**bsdkrun**](https://github.com/tsirysndr/bsdkrun) — a
  Firecracker-style microVM launcher for **BSD and Linux** guests on macOS and
  Linux, built on [libkrun](https://github.com/containers/libkrun).

  The SDK is a thin, stateless wrapper around the `bsdkrun` binary: it builds
  argv, shells out via `System.cmd/3`, and parses JSON output. It has one runtime
  dependency (`:jason`).

      {:ok, box} = Bsdkrun.create(os: :linux, image: "alpine")
      {:ok, res} = Bsdkrun.exec(box, ["uname", "-a"])
      IO.puts(Bsdkrun.Types.Result.text(res))
      :ok = Bsdkrun.stop(box)

  The binary is resolved (in order) from `Bsdkrun.Binary.set_binary_path/1`, the
  `$BSDKRUN_BIN` env var, `bsdkrun` on `$PATH`, or an in-repo dev build. See
  `Bsdkrun.Binary`.

  ## Modules

    * `Bsdkrun.Sandbox`  — create / inspect / drive microVMs.
    * `Bsdkrun.Images`   — list downloaded images.
    * `Bsdkrun.Volumes`  — list / remove persistent volumes.
    * `Bsdkrun.Networks` — global networks (reach machines by name).
    * `Bsdkrun.System`   — probe, fetch BSD images, list versions, grow disks.
    * `Bsdkrun.Types`    — result structs + `Bsdkrun.Types.Result`.
    * `Bsdkrun.Error`    — the single exception type.
  """

  alias Bsdkrun.Sandbox

  @doc "See `Bsdkrun.Sandbox.create/1`."
  defdelegate create(opts), to: Sandbox

  @doc "See `Bsdkrun.Sandbox.create!/1`."
  defdelegate create!(opts), to: Sandbox

  @doc "See `Bsdkrun.Sandbox.get/1`."
  defdelegate get(id), to: Sandbox

  @doc "See `Bsdkrun.Sandbox.list/1`."
  defdelegate list(opts \\ []), to: Sandbox

  @doc "See `Bsdkrun.Sandbox.exec/3`."
  defdelegate exec(ref, command, opts \\ []), to: Sandbox

  @doc "See `Bsdkrun.Sandbox.logs/2`."
  defdelegate logs(ref, opts \\ []), to: Sandbox

  @doc "See `Bsdkrun.Sandbox.stop/1`."
  defdelegate stop(ref), to: Sandbox

  @doc "See `Bsdkrun.Sandbox.start/1`."
  defdelegate start(ref), to: Sandbox

  @doc "See `Bsdkrun.Sandbox.remove/2`."
  defdelegate remove(ref, opts \\ []), to: Sandbox

  @doc "See `Bsdkrun.Binary.set_binary_path/1`."
  defdelegate set_binary_path(path), to: Bsdkrun.Binary
end

File: sdk/ruby/lib/bsdkrun.rb
# frozen_string_literal: true

require_relative "bsdkrun/version"
require_relative "bsdkrun/errors"
require_relative "bsdkrun/binary"
require_relative "bsdkrun/process"
require_relative "bsdkrun/args"
require_relative "bsdkrun/cache"
require_relative "bsdkrun/filesystem"
require_relative "bsdkrun/types"
require_relative "bsdkrun/sandbox"
require_relative "bsdkrun/images"
require_relative "bsdkrun/volumes"
require_relative "bsdkrun/networks"
require_relative "bsdkrun/system"
require_relative "bsdkrun/websocket_frame"
require_relative "bsdkrun/ws_client"
require_relative "bsdkrun/shell_session"
require_relative "bsdkrun/client"

# bsdkrun — a Ruby SDK for {https://github.com/tsirysndr/bsdkrun bsdkrun}, a
# Firecracker-style microVM launcher for BSD, Linux, and unikernel guests. A thin wrapper
# around the +bsdkrun+ CLI: it builds argv, shells out, and parses JSON output.
#
# @example
#   require "bsdkrun"
#
#   sbx = Bsdkrun::Sandbox.create(os: "linux", image: "alpine")
#   puts sbx.exec(["uname", "-a"]).text
#   sbx.stop
module Bsdkrun
  class << self
    # Force the SDK to use a specific +bsdkrun+ binary, bypassing discovery.
    # @param path [String, nil]
    # @return [void]
    def binary_path=(path)
      Binary.override = path
    end

    # @return [String] the resolved +bsdkrun+ binary path.
    # @raise [BinaryNotFound]
    def binary_path
      Binary.resolve
    end

    # Reset cached binary discovery state (mainly for tests).
    # @return [void]
    def reset_binary!
      Binary.reset!
    end

    # Convenience alias for {Sandbox.create}.
    # @return [Sandbox]
    def create(opts = {}, **kwargs)
      Sandbox.create(opts, **kwargs)
    end

    # Convenience alias for {Sandbox.get}.
    # @return [Sandbox]
    def get(id)
      Sandbox.get(id)
    end

    # Convenience alias for {Sandbox.list}.
    # @return [Array<SandboxInfo>]
    def list(all: false)
      Sandbox.list(all: all)
    end

    # The {Images} namespace.
    # @return [Module]
    def images
      Images
    end

    # The {Volumes} namespace.
    # @return [Module]
    def volumes
      Volumes
    end

    # The {Networks} namespace.
    # @return [Module]
    def networks
      Networks
    end

    # The {System} namespace.
    # @return [Module]
    def system
      System
    end
  end
end
